RecoTracker/DeDx/plugins/DeDxHitCalibrator.cc

```diff
@@ -0,0 +1,399 @@
+      pixelCalib_(iConfig, consumesCollector()) {
+  produces<reco::TrackDeDxHitsCollection>("PixelHits");
+  produces<reco::TrackDeDxHitsCollection>("StripHits");
+}
+
+void DeDxHitCalibrator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<bool>("applyGain", true);
+  desc.add<double>("MeVPerElectron", 3.61e-06);
+  desc.add<edm::InputTag>("trackProducer", edm::InputTag("generalTracks"));
+  desc.add<edm::InputTag>("dedxHitInfo", edm::InputTag("dedxHitInfo"));
+  descriptions.add("dedxHitCalibrator", desc);
+}
+
+void DeDxHitCalibrator::beginRun(edm::Run const&, const edm::EventSetup& iSetup) {
+  dedxCalib_ = iSetup.getHandle(dedxCalibToken_);
+  tkGeom_ = iSetup.getHandle(tkGeomToken_);
+}
+
+void DeDxHitCalibrator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& tracks = iEvent.getHandle(tracksToken_);
+  const auto& dedxHitInfo = iEvent.get(dedxHitInfoToken_);
+  pixelCalib_.setESObjects(iSetup);
+
+  // creates the output collection
+  auto pixelTrackDeDxHitAss = std::make_unique<reco::TrackDeDxHitsCollection>(reco::TrackRefProd(tracks));
+  auto stripTrackDeDxHitAss = std::make_unique<reco::TrackDeDxHitsCollection>(reco::TrackRefProd(tracks));
+
+  for (size_t i = 0; i < tracks->size(); i++) {
+    const auto& track = reco::TrackRef(tracks, i);
+    const auto& dedxHits = dedxHitInfo[track];
+    if (dedxHits.isNull())
+      continue;
+    reco::DeDxHitCollection pixelHits, stripHits;
+    processHitInfo(*dedxHits, track->p(), pixelHits, stripHits);
+    pixelTrackDeDxHitAss->setValue(i, pixelHits);
+    stripTrackDeDxHitAss->setValue(i, stripHits);
+  }
+
+  iEvent.put(std::move(pixelTrackDeDxHitAss), "PixelHits");
+  iEvent.put(std::move(stripTrackDeDxHitAss), "StripHits");
+}
+
+/*****************************************************************************/
+void DeDxHitCalibrator::processHitInfo(const reco::DeDxHitInfo& info,
+                                       const float& trackMomentum,
+                                       reco::DeDxHitCollection& pixelHits,
+                                       reco::DeDxHitCollection& stripHits) {
+  for (size_t i = 0; i < info.size(); i++) {
+    // Require hits to be complete and compatible
+    const auto& type = info.type(i);
+    if (!(type & (1 << reco::DeDxHitInfo::Complete)) || !(type & (1 << reco::DeDxHitInfo::Compatible)))
+      continue;
+
+    // Effective path length
+    const auto& pl = info.pathlength(i);
+    const auto pathLength = pl * (1. + a_ * std::log(pl / l0_));
+    const DetId& detId = info.detId(i);
+
+    // Strip
+    if (const auto& stripCluster = info.stripCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const StripGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, stripCluster->barycenter() / 128.);
```

Comment:
`128` could be taken from here:

https://github.com/cms-sw/cmssw/blob/1234e950f3ee35b6f39abdeba60b2d1a53c0c891/DataFormats/SiStripCommon/interface/ConstantsForHardwareSystems.h#L43